frontend_tests/casper_tests/10-admin.js

```diff
@@ -297,9 +297,12 @@ casper.then(function () {
     var selector = 'img#realm-settings-icon[src^="https://secure.gravatar.com/avatar/"]';
     casper.waitUntilVisible(selector, function () {
         casper.test.assertEqual(casper.visible('#realm_icon_delete_button'), false);
+        // A little hack is used her, rather than submitting the form we have just filled
+        // the form and then triggered click event by clicking the button.
         casper.fill('form.admin-realm-form', {
                 realm_icon_file_input: 'static/images/logo/zulip-icon-128x128.png',
-        }, true);
+        }, false);
+        casper.click("#realm_icon_upload_button");
```

Comment:
Now, this is little wacky, I understood the whole method of updating the icon and thought that the above method could work, and really it works. 
Though it is not totally logical, but, somewhat reasonable. I first fill the form for icon and then triggered the click event and it worked. I don't know whether this is easily understandable for future maintainer but I can't get any better solution here.(Also it is still a mystery why the hack in my first comment works).

Reply:
I think it's probably something to do with the fact that we're doing the work inside a click handler, not via the "post" action attached to the form.